File: java/src/org/iota/wasplib/contracts/fairroulette/FairRoulette.java
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package org.iota.wasplib.contracts.fairroulette;

import org.iota.wasplib.client.context.ScCallContext;
import org.iota.wasplib.client.exports.ScExports;
import org.iota.wasplib.client.hashtypes.ScAgent;
import org.iota.wasplib.client.hashtypes.ScColor;
import org.iota.wasplib.client.keys.Key;
import org.iota.wasplib.client.mutable.ScMutableBytesArray;
import org.iota.wasplib.client.mutable.ScMutableMap;

import java.util.ArrayList;

public class FairRoulette {
	private static final Key keyBets = new Key("bets");
	private static final Key keyColor = new Key("color");
	private static final Key keyLastWinningColor = new Key("last_winning_color");
	private static final Key keyLockedBets = new Key("locked_bets");
	private static final Key keyPlayPeriod = new Key("play_period");

	private static final int numColors = 5;
	private static final int defaultPlayPeriod = 120;

	public static void onLoad() {
		ScExports exports = new ScExports();
		exports.AddCall("place_bet", FairRoulette::placeBet);
		exports.AddCall("lock_bets", FairRoulette::lockBets);
		exports.AddCall("pay_winners", FairRoulette::payWinners);
		exports.AddCall("play_period", FairRoulette::playPeriod);
		exports.AddCall("nothing", ScExports::nothing);
	}

	public static void placeBet(ScCallContext sc) {
		long amount = sc.Incoming().Balance(ScColor.IOTA);
		if (amount == 0) {
			sc.Log("Empty bet...");
			return;
		}
		long color = sc.Params().GetInt(keyColor).Value();
		if (color == 0) {
			sc.Log("No color...");
			return;
		}
		if (color < 1 || color > numColors) {
			sc.Log("Invalid color...");
			return;
		}

		BetInfo bet = new BetInfo();
		{
			bet.better = sc.Caller();
			bet.amount = amount;
			bet.color = color;
		}

		ScMutableMap state = sc.State();
		ScMutableBytesArray bets = state.GetBytesArray(keyBets);
		int betNr = bets.Length();
		bets.GetBytes(betNr).SetValue(BetInfo.encode(bet));
		if (betNr == 0) {
			long playPeriod = state.GetInt(keyPlayPeriod).Value();
			if (playPeriod < 10) {
				playPeriod = defaultPlayPeriod;
			}
			sc.Post("lock_bets").Post(playPeriod);
		}
	}

	public static void lockBets(ScCallContext sc) {
		// can only be sent by SC itself
		if (!sc.From(sc.Contract().Id())) {
			sc.Log("Cancel spoofed request");
			return;
		}

		// move all current bets to the locked_bets array
		ScMutableMap state = sc.State();
		ScMutableBytesArray bets = state.GetBytesArray(keyBets);
		ScMutableBytesArray lockedBets = state.GetBytesArray(keyLockedBets);
		int nrBets = bets.Length();
		for (int i = 0; i < nrBets; i++) {
			byte[] bytes = bets.GetBytes(i).Value();
			lockedBets.GetBytes(i).SetValue(bytes);
		}
		bets.Clear();

		sc.Post("pay_winners").Post(0);
	}

	public static void payWinners(ScCallContext sc) {
		// can only be sent by SC itself
		ScAgent scId = sc.Contract().Id();
		if (!sc.From(scId)) {
			sc.Log("Cancel spoofed request");
			return;
		}

		long winningColor = sc.Utility().Random(5) + 1;
		ScMutableMap state = sc.State();
		state.GetInt(keyLastWinningColor).SetValue(winningColor);

		// gather all winners and calculate some totals
		long totalBetAmount = 0;
		long totalWinAmount = 0;
		ScMutableBytesArray lockedBets = state.GetBytesArray(keyLockedBets);
		ArrayList<BetInfo> winners = new ArrayList<BetInfo>();
		int nrBets = lockedBets.Length();
		for (int i = 0; i < nrBets; i++) {
			BetInfo bet = BetInfo.decode(lockedBets.GetBytes(i).Value());
			totalBetAmount += bet.amount;
			if (bet.color == winningColor) {
				totalWinAmount += bet.amount;
				winners.add(bet);
			}
		}
		lockedBets.Clear();

		if (winners.size() == 0) {
			sc.Log("Nobody wins!");
			// compact separate UTXOs into a single one
			sc.Transfer(scId, ScColor.IOTA, totalBetAmount);
			return;
		}

		// pay out the winners proportionally to their bet amount
		int totalPayout = 0;
		int size = winners.size();
		String text;
		for (int i = 0; i < size; i++) {
			BetInfo bet = winners.get(i);
			long payout = totalBetAmount * bet.amount / totalWinAmount;
			if (payout != 0) {
				totalPayout += payout;
				sc.Transfer(bet.better, ScColor.IOTA, payout);
			}
			text = "Pay " + payout + " to " + bet.better;
			sc.Log(text);
		}

		// any truncation left-overs are fair picking for the smart contract
		if (totalPayout != totalBetAmount) {
			long remainder = totalBetAmount - totalPayout;
			text = "Remainder is " + remainder;
			sc.Log(text);
			sc.Transfer(scId, ScColor.IOTA, remainder);
		}
	}

	public static void playPeriod(ScCallContext sc) {
		// can only be sent by SC owner
		if (!sc.From(sc.Contract().Owner())) {
			sc.Log("Cancel spoofed request");
			return;
		}

		long playPeriod = sc.Params().GetInt(keyPlayPeriod).Value();
		if (playPeriod < 10) {
			sc.Log("Invalid play period...");
			return;
		}

		sc.State().GetInt(keyPlayPeriod).SetValue(playPeriod);
	}
}

File: contracts/fairroulette/fairroulette.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package fairroulette

import "github.com/iotaledger/wasplib/client"

const MaxNumber = 5
const DefaultPlayPeriod = 120

func funcPlaceBet(ctx *client.ScCallContext) {
	amount := ctx.Incoming().Balance(client.IOTA)
	if amount == 0 {
		ctx.Panic("Empty bet...")
	}
	number := ctx.Params().GetInt(ParamNumber).Value()
	if number == 0 {
		ctx.Panic("No number...")
	}
	if number < 1 || number > MaxNumber {
		ctx.Panic("Invalid number...")
	}

	bet := &BetInfo{
		Better: ctx.Caller(),
		Amount: amount,
		Number: number,
	}

	state := ctx.State()
	bets := state.GetBytesArray(VarBets)
	betNr := bets.Length()
	bets.GetBytes(betNr).SetValue(EncodeBetInfo(bet))
	if betNr == 0 {
		playPeriod := state.GetInt(VarPlayPeriod).Value()
		if playPeriod < 10 {
			playPeriod = DefaultPlayPeriod
		}
		ctx.Post(&client.PostRequestParams{
			ContractId: ctx.ContractId(),
			Function:   HFuncLockBets,
			Params:     nil,
			Transfer:   nil,
			Delay:      playPeriod,
		})
	}
}

func funcLockBets(ctx *client.ScCallContext) {
	// can only be sent by SC itself
	if !ctx.From(ctx.ContractId().AsAgentId()) {
		ctx.Panic("Cancel spoofed request")
	}

	// move all current bets to the locked_bets array
	state := ctx.State()
	bets := state.GetBytesArray(VarBets)
	lockedBets := state.GetBytesArray(VarLockedBets)
	nrBets := bets.Length()
	for i := int32(0); i < nrBets; i++ {
		bytes := bets.GetBytes(i).Value()
		lockedBets.GetBytes(i).SetValue(bytes)
	}
	bets.Clear()

	ctx.Post(&client.PostRequestParams{
		ContractId: ctx.ContractId(),
		Function:   HFuncPayWinners,
		Params:     nil,
		Transfer:   nil,
		Delay:      0,
	})
}

func funcPayWinners(ctx *client.ScCallContext) {
	// can only be sent by SC itself
	scId := ctx.ContractId().AsAgentId()
	if !ctx.From(scId) {
		ctx.Panic("Cancel spoofed request")
	}

	winningNumber := ctx.Utility().Random(5) + 1
	state := ctx.State()
	state.GetInt(VarLastWinningNumber).SetValue(winningNumber)

	// gather all winners and calculate some totals
	totalBetAmount := int64(0)
	totalWinAmount := int64(0)
	lockedBets := state.GetBytesArray(VarLockedBets)
	winners := make([]*BetInfo, 0)
	nrBets := lockedBets.Length()
	for i := int32(0); i < nrBets; i++ {
		bet := DecodeBetInfo(lockedBets.GetBytes(i).Value())
		totalBetAmount += bet.Amount
		if bet.Number == winningNumber {
			totalWinAmount += bet.Amount
			winners = append(winners, bet)
		}
	}
	lockedBets.Clear()

	if len(winners) == 0 {
		ctx.Log("Nobody wins!")
		// compact separate bet deposit UTXOs into a single one
		ctx.TransferToAddress(scId.Address(), client.NewScTransfer(client.IOTA, totalBetAmount))
		return
	}

	// pay out the winners proportionally to their bet amount
	totalPayout := int64(0)
	size := len(winners)
	for i := 0; i < size; i++ {
		bet := winners[i]
		payout := totalBetAmount * bet.Amount / totalWinAmount
		if payout != 0 {
			totalPayout += payout
			ctx.TransferToAddress(bet.Better.Address(), client.NewScTransfer(client.IOTA, payout))
		}
		text := "Pay " + ctx.Utility().String(payout) +
			" to " + bet.Better.String()
		ctx.Log(text)
	}

	// any truncation left-overs are fair picking for the smart contract
	if totalPayout != totalBetAmount {
		remainder := totalBetAmount - totalPayout
		text := "Remainder is " + ctx.Utility().String(remainder)
		ctx.Log(text)
		ctx.TransferToAddress(scId.Address(), client.NewScTransfer(client.IOTA, remainder))
	}
}

func funcPlayPeriod(ctx *client.ScCallContext) {
	// can only be sent by SC creator
	if !ctx.From(ctx.ContractCreator()) {
		ctx.Panic("Cancel spoofed request")
	}

	playPeriod := ctx.Params().GetInt(ParamPlayPeriod).Value()
	if playPeriod < 10 {
		ctx.Panic("Invalid play period...")
	}

	ctx.State().GetInt(VarPlayPeriod).SetValue(playPeriod)
}
